### PR TITLE
Features

### PR DESCRIPTION
**Pull request by muhenge**
- my_each
- my_each_with_index
- my_select
- my_none?
- my_any?
- my_all?
- my_count
- my_map
- my_inject

### CODE REVIEW

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Good Job :+1: 
Your code looks good. I tested your code and we have some failing edge cases we need to fix before we can merge.
- check my comments for suggestions

### Martin Afani

File: my_enumerable_methods.rb [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    end
+    puts false
+  end
+
+  def my_none?
+    puts true if block_given?
+    my_each do |element|
+      puts false if yield element
+    end
+  end
```
Comment (whiz25) [inline my_enumerable_methods.rb]: - [ ] when a class is passed as an argument returns true if none of the collection is a member of such class
- [ ] should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
array = Array.new(100) { rand(0...9) }
array.my_none?(String) == true #true
```
- [ ] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
```ruby
words = %w[dog door rod blade]
words.my_none?(5) == words.none?(5) #true
```
- [ ] Regex is passed as an argument returns true only if none of the collection matches the Regex
```ruby
 words.my_none?(/z/) == words.none?(/z/) #true
 words.my_none?(/d/) == words.none?(/d/) #true
```
- [ ] returns true if the block never returns true for all elements
```ruby
true_array = [nil, false, true, []]
false_array = [nil, false, nil, false]
false_array.my_none? #true
true_array.my_none? #false
```
- [ ] when no block or argument is given
```ruby
true_array = [nil, false, true, []] 
false_array = [nil, false, nil, false]
false_array.my_none? == true #true
true_array.my_none? == false #false
```

File: my_enumerable_methods.rb [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+  def my_all?
+    my_each do |element|
+      if block_given?
+        yield(element)
+      elsif element.nil? or !i or !yield(element)
+        puts false
+      end
+    end
+    puts true
+  end
```
Comment (whiz25) [inline my_enumerable_methods.rb]: - [ ] should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when no block or argument is given returns true when none of the collection members are false or nil
```ruby
true_array = [1, true, 'hi', []] 
false_array = [1, false, 'hi', []] 
true_array.my_all? == true_array.all? #true
```
- [ ] when a class is passed as an argument returns true if all of the collection is a member of such class
```ruby
array = Array.new(100){rand(0...9)}
array.my_all?(Integer) == array.all?(Integer)
```
- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex
```ruby
words = %w[dog door rod blade]
words.my_all?(/d/) == words.all?(/d/)
```
- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
```ruby
array.my_all?(3) == array.all?(3) #true
```
- [ ] returns true if the block never returns false or nil
```ruby
true_block = proc { |num| num <= 9 } 
false_block = proc { |num| num > 9 } 
array.my_all?(&true_block) == array.all?(&true_block) #true
array.my_all?(&false_block) == array.all?(&false_block) #true
```

File: my_enumerable_methods.rb [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  def my_any?
+    my_each do |element|
+      if block_given?
+        yield element
+      elsif element.nil? or element or yield(element)
+        puts true
+      end
+    end
+    puts false
+  end
```
Comment (whiz25) [inline my_enumerable_methods.rb]: - [ ] should be identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when no block or argument is given returns true if at least one of the collection is not false or nil
```ruby
true_array = [nil, false, true, []]
true_array.my_any? == true_array.any? #true
```
- [ ]  when a class is passed as an argument returns true if at least one of the collection is a member of such class
```ruby
words = %w[dog door rod blade]
words.my_any?(Integer) == words.any?(Integer) #true
```
- [ ] when a Regex is passed as an argument returns false if none of the collection matches the Regex
```ruby
words.my_any?(/z/) == words.any?(/z/) #true
```
- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
```ruby
words.my_any?('cat') == words.any?('cat') #true
```
- [ ] returns true if the block ever returns a value other than false or nil
```ruby
true_block = proc { |num| num <= 9 } 
false_block = proc { |num| num > 9 } 
array.my_any?(&true_block) == array.any?(&true_block) #true
array.my_any?(&false_block) == array.any?(&false_block) #true
```

File: my_enumerable_methods.rb [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    end
+  end
+
+  def my_select
+    select_in = []
+    my_each do |counter|
+      select_in << yield(counter)
+    end
+    select_in
+  end
```
Comment (whiz25) [inline my_enumerable_methods.rb]: - [ ] should be identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] returns an enumerator if no block is given
```ruby
array = Array.new(100){rand(0...9)}
array.my_select == Enumerator #true
```
- [ ] returns an array containing all elements of enum for which the given block returns a true value
```ruby
array = Array.new(100){rand(0...9)}
array.my_select(&block) == array.select(&block) # should return true
```

File: my_enumerable_methods.rb [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+      if block_given?
+        counter += 1 if proc.call(element)
+      elsif !el.nil?
+        counter += 1 if element == arg
+      else
+        count == length
+      end
+    end
+    counter
+  end
```
Comment (whiz25) [inline my_enumerable_methods.rb]: - [ ] should be identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] returns the number of items in enum through enumeration
```ruby
array = Array.new(100){rand(0...9)}
array.my_count == array.count #true
```
- [ ] counts the number of items in enum that are equal to item if an argument is given
```ruby
array.my_count(0) == array.count(0) #true
```

File: my_enumerable_methods.rb [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    if block_given?
+      map = []
+      my_each { |element| mapped << proc.call(element) }
+    elsif proc.nil?
+      my_each { |element| mapped << yield(element) }
+    else
+      enum_for(:map)
+    end
+    map
+  end
```
Comment (whiz25) [inline my_enumerable_methods.rb]: - [ ] should be identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] returns a new array with the results of running block once for every element in enum.
```ruby
array = Array.new(100){rand(0...9)}
array.my_map(&block) == array.map(&block) #true
```
- [ ] returns an Enumerator if no block is given
```ruby
array.my_map == Enumerator # should return true
```

File: my_enumerable_methods.rb [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    return nil unless block_given?
+
+    res = if !elmt.nil?
+            elmt
+          else
+            self[0]
+          end
+    self[0..-1].my_each { |i| res = yield(res, i) }
+    res
+  end
```
Comment (whiz25) [inline my_enumerable_methods.rb]: - [ ] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
array = Array.new(100) { rand(0...9) }
array.my_inject(:+) == array.inject(:+) #true
```
- [ ] combines all elements of enum by applying a binary operation, specified by a block
```ruby
operation = proc { |sum, n| sum + n } 
search = proc { |memo, word| memo.length > word.length ? memo : word } 
array = Array.new(100){rand(0...9)}
words = %w[dog door rod blade]

array.my_inject(&operation) == array.inject(&operation) #true
actual = range.my_inject(4) { |prod, n| prod * n }
expected = range.inject(4) { |prod, n| prod * n }
actual == expected #true
words.my_inject(&search) == words.inject(&search) #true
```

**Review by whiz25 — CHANGES_REQUESTED** [review]

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @muhenge        
#### Great Job so far, but there are a few more things required. 
### General
- [x] `my_any?`:
- returns true if the block ever returns a value other than false or nil
`block = proc { |num| num <= 3 }`
`p [12,12,13,1,4].my_any?(&block)` should return `true`

- when no block or argument is given returns true if at least one of the collection is not false or nil
`p [nil, true, nil].my_any?` - should return `true`

- [x] `my_none?`:
- When no block or argument is given returns true only if none of the collection members is true
`p  [nil, false, true, []].my_none?`  should return `false`

- [x] `my_inject`:
- When a symbol is specified combines each element of the collection by applying the symbol as a named method
`p (5..10).my_inject(2, :*) - should return 302400`

### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Comment by Dandush03** [comment]
### :heavy_check_mark: Project Approve :exploding_head: :checkered_flag:
# Hey :wave: 
### Great Job so Far :grey_exclamation: :grey_exclamation: :muscle: :smiley: 

### Following Previous Reviews and Seeing you Have Made Required Changes. 
### The project Looks Great.:grey_exclamation::wink:.

### Good Luck in your Next Assignment
### Happy coding:computer: 
#### Daniel Laloush

**Pull request merged**